java-checks/src/main/java/org/sonar/java/checks/ForLoopVariableTypeCheck.java

```diff
@@ -0,0 +1,118 @@
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.ForEachStatement;
+import org.sonar.plugins.java.api.tree.Tree;
+
+@Rule(key = "S4838")
+public class ForLoopVariableTypeCheck extends IssuableSubscriptionVisitor {
+
+  private static final String PRIMARY_MESSAGE = "Change \"%s\" to the type handled by the Collection.";
+  private static final String SECONDARY_MESSAGE = "Collection item type is \"%s\"";
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.FOR_EACH_STATEMENT);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+
+    ForEachStatement actualStatement = (ForEachStatement) tree;
+    Type variableType = actualStatement.variable().type().symbolType();
+    Type collectionItemType = getCollectionItemType(actualStatement.expression());
+
+    if (collectionItemType != null && !isMostPreciseType(variableType, collectionItemType)) {
+      // Second pass: check if the variable is down-cast in the statement block
+      DownCastVisitor downCastVisitor = new DownCastVisitor(actualStatement, variableType, collectionItemType);
+      downCastVisitor.setContext(context);
+      downCastVisitor.scanTree(((ForEachStatement) tree).statement());
+    }
+  }
+
+  @CheckForNull
+  private static JavaType getCollectionItemType(ExpressionTree expression) {
+    if (((JavaType) expression.symbolType()).isParameterized()) {
+      ParametrizedTypeJavaType paramType = (ParametrizedTypeJavaType) expression.symbolType();
+      return paramType.substitution(paramType.typeParameters().get(0));
+    } else {
+      return null;
+    }
+  }
+
+  private static boolean isMostPreciseType(Type variableType, Type collectionItemType) {
+    if (collectionItemType instanceof WildCardType) {
+      return ((WildCardType) collectionItemType).isSubtypeOfBound((JavaType) variableType);
+    } else if (collectionItemType instanceof ParametrizedTypeJavaType) {
+      return ((ParametrizedTypeJavaType) collectionItemType).erasure().equals(variableType.erasure());
+    } else {
+      return variableType.equals(collectionItemType);
+    }
+  }
+
+  private class DownCastVisitor extends SubscriptionVisitor {
+
+    private final ForEachStatement forEachStatement;
+    private final Type variableType;
+    private final Type collectionItemType;
+
+    private DownCastVisitor(ForEachStatement forEachStatement, Type variableType, Type collectionItemType) {
+      this.forEachStatement = forEachStatement;
+      this.variableType = variableType;
+      this.collectionItemType = collectionItemType;
+    }
+
+    @Override
+    public List<Tree.Kind> nodesToVisit() {
+      return Collections.singletonList(Tree.Kind.TYPE_CAST);
+    }
+
+    @Override
+    public void visitNode(Tree tree) {
```

Comment:
You should check that the cast is related to the variable. Otherwise any unrelated cast would raise the issue.
Ex:
```
J obj;
    for(I i: collectionOfK) { // Noncompliant
      (L) obj;
    }
```